### PR TITLE
[WIP] fix checkpointing on TPU

### PR DESCRIPTION
**Pull request by matt-peters**
## What does this PR do?

Draft PR: fixes checkpointing on TPU, see #2700 

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -266,12 +266,12 @@ def on_train_start(self, trainer, pl_module):
         trainer.ckpt_path = ckpt_path
         trainer.weights_save_path = ckpt_path
 
-    @rank_zero_only
     def on_validation_end(self, trainer, pl_module):
-        # only run on main process
-        if trainer.global_rank != 0:
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:272]: This may break DDp

File: pytorch_lightning/trainer/training_io.py [hunk 2]
```diff
@@ -261,22 +261,36 @@ def _atomic_save(self, checkpoint, filepath: str):
                 This points to the file that the checkpoint will be stored in.
         """
         tmp_path = str(filepath) + ".part"
-        torch.save(checkpoint, tmp_path)
-        os.replace(tmp_path, filepath)
+        if self.use_tpu:
```
Comment (Borda) [inline pytorch_lightning/trainer/training_io.py:264]: There is `on_tpu`

File: pytorch_lightning/trainer/training_io.py [hunk 3]
```diff
@@ -261,22 +261,36 @@ def _atomic_save(self, checkpoint, filepath: str):
...
-                self._atomic_save(checkpoint, filepath)
+                self._atomic_save(chkpt, filepath)
+
+        checkpoint = self.dump_checkpoint(weights_only)
+
+        # self._atomic_save has different behavior for XLA vs
+        # non-XLA.  In XLA, it has a barrier and internal logic to only
+        # save for rank==0, so need to call for all ranks. For non-XLA,
+        # it doesn't have rank==0 logic so only call for rank==0
+        if self.use_tpu:
```
Comment (Borda) [inline pytorch_lightning/trainer/training_io.py:290]: Why not or?

**Comment by lezwon** [comment]
@matt-peters I think we will require an `on_tpu` check [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/26afcaa30e109b728aa4753c58c78cee997b3762/pytorch_lightning/trainer/distrib_data_parallel.py#L618) too.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -266,12 +266,12 @@ def on_train_start(self, trainer, pl_module):
...
-            return
-
+        # To get checkpointing working on TPU, need to call _save_model
+        # for all ranks, to avoid deadlocks.  Assuming save_function is mapped
```
Comment (ibeltagy) [inline pytorch_lightning/callbacks/model_checkpoint.py:271]: nit: without `if trainer.global_rank != 0:` and `@rank_zero_only`, all threads are writing to the log

File: pytorch_lightning/trainer/training_io.py [hunk 5]
```diff
@@ -261,22 +261,36 @@ def _atomic_save(self, checkpoint, filepath: str):
                 This points to the file that the checkpoint will be stored in.
         """
         tmp_path = str(filepath) + ".part"
-        torch.save(checkpoint, tmp_path)
-        os.replace(tmp_path, filepath)
+        if self.use_tpu:
+            xm.save(checkpoint, tmp_path, master_only=True, global_master=True)
```
Comment (lezwon) [inline pytorch_lightning/trainer/training_io.py:265]: Maybe add a barrier before `xm.save`, to make sure all processes are in sync?

Comment (ibeltagy) [reply pytorch_lightning/trainer/training_io.py:265]: the barrier is already inside `xm.save` [here](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L814)

**Comment by matt-peters** [comment]
Thanks for the reviews and comments.  I don't have the bandwidth to shepherd this to completion, this PR was meant to outline a fix, not be ready to merge.  I also haven't actually run the code as I'm working off a different release and this was patched while resolving conflicts so caveat emptor...

File: pytorch_lightning/trainer/training_io.py (hunk 5 above)
Comment (lezwon) [reply pytorch_lightning/trainer/training_io.py:265]: Excuse my naiveness, but shouldn't there be one before save, to make sure that the weights have been updated by every process?

**Comment by Borda** [comment]
@matt-peters mind allow editing this pr b maintainers?

**Comment by matt-peters** [comment]
Edits are now allowed.

**Comment by Borda** [comment]
@matt-peters what is missing, is it still wip?

**Comment by matt-peters** [comment]
LGTM

**Comment by Borda** [comment]
it seems that the error comes from Horovod, even this PR shall not touch it... @tgaddair ?

**Comment by tgaddair** [comment]
> it seems that the error comes from Horovod, even this PR shall not touch it... @tgaddair ?

I'm not sure they are Horovod specific, looks like there are similar failures for DDP.

**Comment by edenlightning** [comment]
we finished our refactoring. @matt-peters can you please rebase this change? we would love to get this supported!!

**Comment by Borda** [comment]
@matt-peters how is it going here? mind finish it?

**Comment by matt-peters** [comment]
It looks like `pytorch_lightning/trainer/training_io.py` was removed from `PyTorchLightning:master` at some point, and the functionality presumably moved elsewhere. As most of my changes were in that file, I don't have the bandwidth to track them down, port them over, and test it out.  With write access you should be resolve the conflicts and push to this branch.

**Comment by williamFalcon** [comment]
i can look at it :)

**Comment by Borda** [comment]
@matt-peters how about this one, is it still WIP or ready to review? 🐰 
if so, mind rebase on master and resolve conflicts...

**Comment by Borda** [comment]
@matt-peters maybe after the refactoring it would be easier to reset this to master and make the changes again...
you can just:
```
git reset upstream/master
# do your changes here
git commit ...
git push -f
```

**Comment by Borda** [comment]
@matt-peters @lezwon can we finish this one?

**Comment by lezwon** [comment]
@Borda Have raised a new PR for this issue here: #4309

**Pull request closed without merging**